fetchai/ledger/crypto/deed.py

```diff
@@ -0,0 +1,114 @@
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+import logging
+from enum import Enum
+from typing import Union
+
+from fetchai.ledger.crypto import Address, Identity, Entity
+
+AddressLike = Union[Address, Identity]
+
+
+class InvalidDeedError(Exception):
+    pass
+
+
+class Operation(Enum):
+    """Enables future amendments to the deed"""
+    amend = 1
+    """Enables FET transfers"""
+    transfer = 2
+    """Enab"""
+    execute = 3
+
+    stake = 4
+
+    def __repr__(self):
+        return '<%s.%s>' % (self.__class__.__name__, self.name)
+
+    def __str__(self):
+        return self.name
+
+
+class Deed:
+    def __init__(self, address: AddressLike):
+        self._address = address
+        self._signees = {}
+        self._thresholds = {}
+
+    def set_signee(self, signee: Entity, voting_weight: int):
+        self._signees[signee] = int(voting_weight)
+
+    def remove_signee(self, signee: Entity):
+        if signee in self._signees:
+            del self._signees[signee]
+
+    def set_threshold(self, operation: Operation, threshold: int):
+        if threshold > self.total_votes:
+            raise InvalidDeedError("Attempting to set threshold higher than available votes - it will never be met")
```

Comment:
maybe more of a general point but throwing exceptions on validation is considered bad practise https://stackoverflow.com/questions/1504302/is-it-a-good-or-bad-idea-throwing-exceptions-when-validating-data